### PR TITLE
http status api

### PR DESCRIPTION
**Pull request by aastein**
fixes: #254 

`http://localhost:3001/lease/d6f4b6728c7deb187a07afe8e145e214c716e287039a204e7fac1fc121dc0cef/1/2/8224e14f903a2e136a6362527b19f11935197175cb69981940933aa04459a2a9`
image

`http://localhost:3001/lease/d6f4b6728c7deb187a07afe8e145e214c716e287039a204e7fac1fc121dc0cef/1/2/8224e14f903a2e136a6362527b19f11935197175cb69981940933aa04459a2a9/web`
image

### CODE REVIEW

File: provider/http/server.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+	statusPath        = statusPathPrefix + leaseID
+	serviceStatusPath = statusPathPrefix + leaseID + "/{" + name + "}"
+)
+
+func errorResponse(w http.ResponseWriter, log log.Logger, status int, message string) {
+	log.Error("error", "status", status, "message", message)
+	w.Header().Set("X-Content-Type-Options", "nosniff")
+	w.Header().Set("Content-Type", contentType)
+	w.WriteHeader(status)
+	json.NewEncoder(w).Encode(map[string]string{"message": message})
```
Comment (golangci) [inline provider/http/server.go]: Error return value of `(*encoding/json.Encoder).Encode` is not checked

File: provider/http/server.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+		}
+		if deployments == nil {
+			errorResponse(w, log, http.StatusBadRequest, "no deployments found for lease")
+			return
+		}
+		response := make(map[string]string)
+		for _, deployment := range deployments.Items {
+			response[deployment.Name] = fmt.Sprintf("available replicas: %v/%v", deployment.Status.AvailableReplicas, deployment.Status.Replicas)
+		}
+		json.NewEncoder(w).Encode(response)
```
Comment (golangci) [inline provider/http/server.go]: Error return value of `(*encoding/json.Encoder).Encode` is not checked

File: provider/http/server.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+			errorResponse(w, log, http.StatusBadRequest, "service not found for lease")
+			return
+		}
+		if deployment == nil {
+			errorResponse(w, log, http.StatusBadRequest, "service not found for lease")
+			return
+		}
+		status := deployment.Status
+		status.Conditions = nil
+		json.NewEncoder(w).Encode(status)
```
Comment (golangci) [inline provider/http/server.go]: Error return value of `(*encoding/json.Encoder).Encode` is not checked

File: provider/http/server.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	donech := make(chan struct{})
+
+	go func() {
+		defer close(donech)
+		<-ctx.Done()
+		log.Info("Shutting down server")
+		server.Shutdown(context.Background())
```
Comment (golangci) [inline provider/http/server.go]: Error return value of `server.Shutdown` is not checked

File: types/id.go [hunk 5]
```diff
@@ -170,3 +170,20 @@ func (id LeaseID) GroupID() DeploymentGroupID {
 func (id LeaseID) DeploymentID() base.Bytes {
 	return id.Deployment
 }
+
+func NewLeaseID(deployment, group, order, provider string) (LeaseID, error) {
```
Comment (boz) [inline types/id.go]: keep this private in http.  edit: nvm looks like it's not used, delete it.

File: provider/http/server_test.go [hunk 6]
```diff
@@ -63,7 +55,8 @@ func withServer(t *testing.T, fn func(), h *pmanifest.Handler) {
 
 	go func() {
 		defer close(donech)
-		err := RunServer(ctx, testutil.Logger(), "3001", h)
+		err := RunServer(ctx, testutil.Logger(), "3001", h, c)
+		fmt.Println("server exited err:", err)
```
Comment (boz) [inline provider/http/server_test.go]: ^

File: provider/http/server_test.go [hunk 7]
```diff
@@ -35,26 +38,15 @@ func TestManifest(t *testing.T) {
...
-		err = Send(mani, signer, provider, deployment)
-		require.NoError(t, err)
-	}, handler)
-}
-
-func TestStatus(t *testing.T) {
-	handler := new(pmock.Handler)
-
-	withServer(t, func() {
-		resp, err := http.Get("http://localhost:3001/status")
```
Comment (boz) [inline provider/http/server_test.go]: Let's continue to test this.  I know `status` is overloaded, maybe call it `healthcheck`.  It's used in `_run/multi`

File: provider/cluster/client.go [hunk 8]
```diff
@@ -1,12 +1,17 @@
...
+	"k8s.io/api/apps/v1"
+)
 
 type Client interface {
 	Deploy(types.LeaseID, *types.ManifestGroup) error
 	Teardown(types.LeaseID) error
 
 	Deployments() ([]Deployment, error)
+	KubeDeployments(types.LeaseID) (*v1.DeploymentList, error)
+	KubeDeployment(types.LeaseID, string) (*v1.Deployment, error)
```
Comment (boz) [inline provider/cluster/client.go]: I'd prefer to keep this ignorant of kube -- make a simplified, unified status struct -- but for now it's ok.

File: types/id.go [hunk 9]
```diff
@@ -169,4 +169,4 @@ func (id LeaseID) GroupID() DeploymentGroupID {
 
 func (id LeaseID) DeploymentID() base.Bytes {
 	return id.Deployment
-}
+}
```
Comment (golangci) [inline types/id.go]: File is not goimports-ed

File: provider/http/server.go [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+			response[deployment.Name] = fmt.Sprintf("available replicas: %v/%v", deployment.Status.AvailableReplicas, deployment.Status.Replicas)
+		}
+		json.NewEncoder(w).Encode(response)
+	}
+}
+
+func newStatusHandler() func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "text/plain; charset=us-ascii")
+		w.Write([]byte("OK\n"))
```
Comment (golangci) [inline provider/http/server.go]: Error return value of `w.Write` is not checked

**Pull request merged**